---
title: "MiDAS"
author: "Christian Hammer & Maciej MigdaÅ‚ & Dan Fu Ruan & others"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

devtools::load_all()
library("dplyr")
library("knitr")
library("kableExtra")
```

# Introduction
`MiDAS` package provide tools for working with HLA data generated by HLA 
inference tools for WGS data such as "HLA-HD". Main purpose of those tools is to
make it as easy as possible to import, transform and analyse HLA data. Moreover 
package also provides functionality to integrate HLA alleles with KIR haplotype 
data.

# Quick start

## Reading input data

```{r loading_data}
# let's load phenotypic data and covariates
pheno_file <- system.file("extdata", "pheno_example.txt", package = "MiDAS")
pheno <- read.table(pheno_file, header = TRUE, stringsAsFactors = FALSE)

kable(pheno) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")

covar_file <- system.file("extdata", "covar_example.txt", package = "MiDAS")
covar <- read.table(covar_file, header = TRUE, stringsAsFactors = FALSE)

kable(covar) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")

# hla calls can be loaded using readHlaCalls function
hla_calls_file <- system.file("extdata", "HLAHD_output_example.txt", package = "MiDAS")
hla_calls <- readHlaCalls(hla_calls_file)

kable(hla_calls) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")

coldata <- dplyr::left_join(pheno, covar, by = "ID")
```

## Creating midas data sets
First step to analyse HLA data is to create a data set that will include all the 
information we will use for statistical modeling. For that we `MiDAS` provides 
`prepareMiDAS` function. In most basic case it will simply join our `hla_calls` 
with `pheno` and `covar` data frames. It can however also be used to transform 
our HLA calls into another type of variable such as amino acids residues or HLA 
alleles supertypes. This can be controled using `analysis_type` argument.

```{r creating_midas_data_set}
midas_data <- prepareMiDAS(hla_calls = hla_calls,
                               colData = coldata,
                               analysis_type = "hla_allele",
                               inheritance_model = "additive"
)

kable(midas_data) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")
```
## Association analysis

```{r association_analysis}
# Logistic regression
object <- glm(OS_DIED ~ AGE + SEX + term, data = midas_data, family = binomial(link = "logit"))
glm_results <- runMiDAS(object, mode = "linear", analysis_type = "hla_allele")

kable(glm_results) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")

# Linear regression
object <- lm(OS ~ AGE + SEX + term, data = midas_data)
lm_results <-  runMiDAS(object, mode = "linear", analysis_type = "hla_allele")

kable(lm_results) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")

# Cox survival analysis
library("survival")
object <- coxph(Surv(OS, OS_DIED) ~ AGE + SEX + term, data = midas_data)
coxph_results <- runMiDAS(object, mode = "linear", analysis_type = "hla_allele")

kable(lm_results) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")
```

# Converting HLA alleles to variables
HLA alleles can be converted to other variables. MiDAS provides conversions to 
types such as amino acids, G groups, Bw groups, C groups, supertypes or 
expression levels. The easiest way to get things going is by using 
`prepareMiDAS`. Conviniently we can convert HLA alleles to all of those 
variables in one go.

```{r hla_to_variables}
midas_data <- prepareMiDAS(hla_calls = hla_calls,
                               colData = coldata,
                               analysis_type = c("hla_allele", 
                                                 "aa_level", 
                                                 "allele_g_group", 
                                                 "allele_supertype", 
                                                 "allele_group"
                               ),
                               inheritance_model = "additive"
)

kable(midas_data) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")

object <- coxph(Surv(OS, OS_DIED) ~ AGE + SEX + term, data = midas_data)
coxph_results <- runMiDAS(object, mode = "linear", analysis_type = "hla_allele", conditional = TRUE)

kable(coxph_results) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")
```

For some applications it might be usefull to convert `hla_calls` to variables
directly (not to counts). This can be acomplished using `hlaToVariable`. This
functions uses dictionaries that are shipped with the package, those can be 
listed with `listMiDASDictionaries`.

```{r hla_to_variable}
listMiDASDictionaries()

supertypes <- hlaToVariable(hla_calls, dictionary = "allele_HLA_supertype")

kable(supertypes) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) %>%
  scroll_box(width = "100%", height = "200px")
```


# Accessing published alleles frequencies
Comming soon...

Right now we are directly querying allelefrequencies.net data base
